ml-prims/src/matrix/math.h

```diff
@@ -273,64 +276,79 @@ void reciprocal(math_t *in, math_t *out, int len) {
  * @{
  */
 
-// TODO: Check with Thejaswi if he can come up with faster approach to this
-// function.
 template <typename math_t>
 void ratio(math_t *src, math_t *dest, int len) {
-  auto counting = thrust::make_counting_iterator(0);
   auto d_src = src;
-  auto s = len;
   auto d_dest = dest;
 
-  thrust::for_each(counting, counting + len, [=] __device__(int idx) {
-    math_t total = 0.0;
-    for (int i = 0; i < s; i++) {
-      total += d_src[i];
-    }
-    if (total != 0.0) {
-      d_dest[idx] = d_src[idx] / total;
-    }
-  });
+  math_t* d_sum;
+  allocate(d_sum, 1);
```

Comment:
Please do not use explicit `allocate` and `cudaFree` calls. Use `DeviceAllocator` object as being used here, for eg: https://github.com/rapidsai/cuml/blob/branch-0.6/ml-prims/src/linalg/svd.h#L148. Basically, have this method accept a DeviceAllocator object and use that for allocation of the required memory.